### DISCUSSION TITLE
Help with PWM - Pin weirdness after calling PWM() using imports?

### DISCUSSION THREAD

**Discussion by jouellnyc** [micropython/micropython#15938] [Hardware & Peripherals]
Good day MP'ers

Putting a small toy car together, using a l9110 motor control and simple motor using  `MicroPython v1.23.0 on 2024-06-02; Raspberry Pi Pico with RP2040`

It seems like the value() of the PIN gets set to 0 zero when creating an instance PWM when using imports but not when directly creating the instance?.  I am having trouble tracking the behavior down.  It also seems I cannot control it or set it back once instantiated? I must be doing something wrong?  I have tried different pins just for contrast.

Snipping non-needed text:  

```
# wheels.py
import time
from machine import Pin, PWM

RIGHT_B_B1A = Pin(15, Pin.OUT)
RIGHT_B_B1A.value(1)
RIGHT_B_B1B = Pin(13, Pin.OUT)
RIGHT_B_B1B.value(1)
print('0 RIGHT_B_B1A ', RIGHT_B_B1A.value())
print('0 RIGHT_B_B1B ', RIGHT_B_B1B.value())

```

```
# drive_api.py
import time
from machine import Pin, PWM
from wheels import RIGHT_B_B1A, RIGHT_B_B1B

class CarControl:
    def __init__(self, speed=65535, freq=1000):
        
        # Right wheel setup
        self.RIGHT_B_B1A = RIGHT_B_B1A
        self.RIGHT_B_B1B = RIGHT_B_B1B
        print('0b  RIGHT_B_B1A ', RIGHT_B_B1A.value())
        print('0b  RIGHT_B_B1B ', RIGHT_B_B1B.value())
        print('1 RIGHT_B_B1A ', self.RIGHT_B_B1A.value())
        print('1 RIGHT_B_B1B ', self.RIGHT_B_B1B.value())
        
        self.Rforward  = PWM(self.RIGHT_B_B1B, freq=freq)
        self.Rbackward = PWM(self.RIGHT_B_B1A, freq=freq)
        print('2 RIGHT_B_B1A ', self.RIGHT_B_B1A.value())
        print('2 RIGHT_B_B1B ', self.RIGHT_B_B1B.value())
   
        self.RIGHT_B_B1A.value(1)
        self.RIGHT_B_B1B.value(1)
        print('3 RIGHT_B_B1A ', self.RIGHT_B_B1A.value())
        print('3 RIGHT_B_B1B ', self.RIGHT_B_B1B.value())

```                

Output with imports:

```
>>> import time
>>> from drive_api import CarControl
>>> percent=1 #1, .9, .8, etc.
>>> car = CarControl(speed=int(65535 * percent), freq=1000)
0 RIGHT_B_B1A  1
0 RIGHT_B_B1B  1
0b  RIGHT_B_B1A  1
0b  RIGHT_B_B1B  1
1 RIGHT_B_B1A  1
1 RIGHT_B_B1B  1
2 RIGHT_B_B1A  0      <------- change to 0  ?
2 RIGHT_B_B1B  0     <------- change to 0   ?
3 RIGHT_B_B1A  0     <------- shouldn't this at least change to 1? 
3 RIGHT_B_B1B  0     <------- shouldn't this at least change to 1? 
```

On the Repl directly I do not see this behavior:

````
from wheels import RIGHT_B_B1A, RIGHT_B_B1B
print('0 RIGHT_B_B1A ', RIGHT_B_B1A.value())
print('0 RIGHT_B_B1B ', RIGHT_B_B1B.value())
Rforward  = PWM(RIGHT_B_B1B, freq=1000)
Rbackward = PWM(RIGHT_B_B1A, freq=1000)
print('1 RIGHT_B_B1A ', RIGHT_B_B1A.value())
print('1 RIGHT_B_B1B ', RIGHT_B_B1B.value())

0 RIGHT_B_B1A  0
0 RIGHT_B_B1B  0
1 RIGHT_B_B1A  0
1 RIGHT_B_B1B  0

```

**Comment by peterhinch** [comment]
When a `Pin` instance has been assigned to a `PWM`, is it valid to access `pin.value()`? I'm not sure what reading the value would be expected to return. What would writing the value be expected to achieve?

To turn a PWM pin on, set duty to 100%. To turn it off, set duty to 0.

**Reply by jouellnyc** [reply]
Thank you. The whole reason I even noticed what I thought was odd was just creating an instance of `CarControl` makes the wheels start to move, so I was trying to track why that was happening down. Since the ``__init__`` method does not fiddle w/the values at all, I was just trying to prove to myself the values did or did not change.

**Reply by peterhinch** [reply]
I suggest you remove all references to `.value` which occur after the PWM is assigned and try again.

**Reply by robert-hh** [reply]
The PWM class at ESP32 has the habit, that with instantiation already an output signal is generated. To avoid that you have so set the duty cycle in the constructor to 0.

**Reply by jouellnyc** [reply]
Thanks @peterhinch and @peterhinch 

It appears that the truth table for the l9110 is off for me for Motor B (B-1A and B-1B).

The motor should be stopped w/(0,0) or (1,1)

https://docs.sunfounder.com/projects/3in1-kit-v2/en/latest/components/component_l9110_module.html

I simplified by only using 1 motor controller and shortening the variable names.


```
"""  With the Two wheels pointing forward - i.e Front Wheel Drive 

#!! Hold the L9110 this way !!

||Right Wheel    || Left Wheel    ||
|| (OB) (OA) || (OB) (OA)          ||
||   Green Term Blocks  x4        ||
|| MOTOR B       ||   MOTOR A  ||
((...............PCB ..............))
| B-1A |B-1B |GND |VCC |A-1A |A-1B |  

### Truth Table as observed 

# STOPPED
>>> B_B1B.value(1); B_B1A.value(1)
>>> B_B1B.value(0); B_B1A.value(1)

# FORWARD
>>> B_B1B.value(1); B_B1A.value(0)
>>> B_B1B.value(0); B_B1A.value(0)

# STOPPED
>>> A_A1A.value(0); A_A1B.value(0)
>>> A_A1A.value(1); A_A1B.value(1)

# FORWARD
>>> A_A1A.value(1); A_A1B.value(0)
# REVERSE
>>> A_A1A.value(0); A_A1B.value(1)

```

I.E Just creating pins w/o pwm makes the wheel move (makes sense knowing the above but not expected per the docs)

I tried 3 L9110's all with the same behavior. 

```
import time
from machine import Pin
B_B1A = Pin(15, Pin.OUT)
B_B1B = Pin(13, Pin.OUT)
A_A1A = Pin(0, Pin.OUT)
A_A1B = Pin(1, Pin.OUT)

```

I'll have to play around with this. Thanks

**Reply by jouellnyc** [reply]
Coming back to this. I took the following steps and made the following observations:

1. I broke down the whole circuit and rebuilt it
2. I tried again w/the first l9110  and got magic smoke - nothing out of the ordinary
3. I tried again w/the second l9110  and got magic smoke - nothing out of the ordinary
4. I tried again w/the third l9110  and all is well - external USB power 5v or even pin#40 on the pico powering it. 
    Forward , backward, 'harsh stops', you name it. It acted fine and none of this weird import related problems.

I have seen this "truth table oddness" while I wired up a L298N, when GND was not wired properly. Maybe that started this whole mess? I thought I was pretty thorough. Or maybe the 2 controllers were just on their way out? Not sure.

I also tried a L298N - no problems whatsoever. I bought a few more l9110's - once I get them I can try them but - as best I can see - it appears to have been related to bad hardware - Or again something I did but cannot quite put my finger on it.

(FWIW, I had the car on the floor basically "done" and working w/an IR remote for a few days  and just wanted to add 'one more quick thing'  and then all this sort of stuff started happening w/motors that looked liked [these](https://www.amazon.com/AEDIKO-Motor-Gearbox-200RPM-Ratio/dp/B09N6NXP4H/)). There ought to be a law...

As before, thanks again for all the replies.
